source/d1celtileset.cpp

```diff
@@ -50,47 +50,67 @@ D1CEL_FRAME_TYPE guessFrameType(QByteArray &rawFrameData)
 
 bool D1CelTileset::load(D1Gfx &gfx, std::map<unsigned, D1CEL_FRAME_TYPE> &celFrameTypes, QString filePath, const OpenAsParam &params)
 {
-    // Opening CEL file with a QBuffer to load it in RAM
-    if (!QFile::exists(filePath))
-        return false;
-
-    QFile file = QFile(filePath);
-
-    if (!file.open(QIODevice::ReadOnly))
-        return false;
+    // prepare file data source
+    QFile file;
+    // done by the caller
+    // if (!params.celFilePath.isEmpty()) {
+    //    filePath = params.celFilePath;
+    // }
+    if (!filePath.isEmpty()) {
+        file.setFileName(filePath);
+        if (!file.open(QIODevice::ReadOnly)) {
+            return false;
+        }
+    }
 
     QByteArray fileData = file.readAll();
     QBuffer fileBuffer(&fileData);
 
-    if (!fileBuffer.open(QIODevice::ReadOnly))
+    if (!fileBuffer.open(QIODevice::ReadOnly)) {
         return false;
+    }
 
     // Read CEL binary data
     QDataStream in(&fileBuffer);
     in.setByteOrder(QDataStream::LittleEndian);
 
-    // CEL HEADER CHECKS
+    // File size check
+    int numFrames = 0;
+    auto fileSize = fileBuffer.size();
+    if (fileSize != 0) {
+        // CEL HEADER CHECKS
+        if (fileSize < 4) {
+            qDebug() << "Invalid level-cel-file I.";
+            return false;
+        }
 
-    // Read first DWORD
-    quint32 firstDword;
-    in >> firstDword;
+        // Read first DWORD
+        quint32 readDword;
+        in >> readDword;
 
-    // Trying to find file size in CEL header
-    if (fileBuffer.size() < (4 + firstDword * 4 + 4))
-        return false;
+        numFrames = readDword;
+
+        // Trying to find file size in CEL header
+        if (fileSize < (4 + numFrames * 4 + 4)) {
+            qDebug() << "Invalid level-cel-file II.";
+            return false;
+        }
 
-    fileBuffer.seek(firstDword * 4 + 4);
-    quint32 fileSizeDword;
-    in >> fileSizeDword;
+        fileBuffer.seek(numFrames * 4 + 4);
+        quint32 fileSizeDword;
+        in >> fileSizeDword;
 
-    gfx.groupFrameIndices.clear();
+        if (fileSize != fileSizeDword) {
+            qDebug() << "Invalid level-cel-file III.";
```

Comment:
```suggestion
            qDebug() << "Unexpected end of level-cel-file.";
```

Reply:
that is not the end of the level-cel file. The header is wrong.